### ISSUE TITLE
Can't install PostGIS extension under Postgres

### ISSUE THREAD

**Issue by sandrasi** [Homebrew/linuxbrew-core#15275]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] are reporting a bug others will be able to reproduce and not asking a question or requesting software. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh. To get software added or changed in Homebrew please file a [Pull Request](https://github.com/Homebrew/linuxbrew-core/blob/master/CONTRIBUTING.md)
- [ ] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, Homebrew/homebrew-core formula (not cask) on macOS? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new/choose. If it's a Linux problem please file this issue at https://github.com/Homebrew/linuxbrew-core/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:

- What you were trying to do (and why)

Have a fresh install of Linuxbrew and I ran `$brew install postgis`. After everything got installed, I started the PostgreSQL server and tried to install the PostGIS extension from the postgres console.

- What happened (include command output)

The postgres console displays an error message:

_"ERROR:  could not open extension control file "/home/linuxbrew/.linuxbrew/share/postgresql/extension/postgis.control": No such file or directory"_

Note, that there is not a trace of this file or anything similar anywhere on the file system. I ran the
```bash
$ sudo find / -iname "*postgis.control*"
``` 
command which returned no results.

- What you expected to happen

The extension gets installed.

- Step-by-step reproduction instructions (by running `brew install` commands)

1.) Have a fresh install of Linuxbrew and follow the post-installation instructions (install build-essentials, add it to the ~/.profile file, and force-update the formulas).
2.) Run the following commands:
```
$ brew install postgis
$ pg_ctl -D /home/linuxbrew/.linuxbrew/var/postgres start
$ psql postgres
```
3.) From the postgres console run the following command:
```
postgres=# CREATE EXTENSION postgis;
```

I've also tried `$ brew reinstall postis` which printed the following logs
```
==> Downloading https://download.osgeo.org/postgis/source/postgis-2.5.3.tar.gz
Already downloaded: /home/sandrasi/.cache/Homebrew/downloads/bb221060ce3d64d32854a7ca9a65cbbe770c0295584e7ac13082343149ced0eb--postgis-2.5.3.tar.gz
==> ./configure --with-projdir=/home/linuxbrew/.linuxbrew/opt/proj --with-jsondir=/home/linuxbrew/.linuxbrew/opt/json-c --with-pgconfig=/home/linuxbrew/.linuxbrew/
==> make
==> make install DESTDIR=/tmp/postgis-20190906-11061-10nmsqm/postgis-2.5.3/stage
Warning: tried to install empty array to /home/linuxbrew/.linuxbrew/Cellar/postgis/2.5.3/bin
Warning: tried to install empty array to /home/linuxbrew/.linuxbrew/Cellar/postgis/2.5.3/share/doc/postgis/postgresql/extension
Warning: tried to install empty array to /home/linuxbrew/.linuxbrew/Cellar/postgis/2.5.3/share/postgresql/extension
Warning: tried to install empty array to /home/linuxbrew/.linuxbrew/Cellar/postgis/2.5.3/share/postgis
Warning: tried to install empty array to /home/linuxbrew/.linuxbrew/Cellar/postgis/2.5.3/share/postgis_topology
🍺  /home/linuxbrew/.linuxbrew/Cellar/postgis/2.5.3: 24 files, 3.8MB, built in 43 seconds
```

Notice the _Warning_ lines above - I wonder what the "empty array" may be.

I've read somewhere that the packaging of PostGIS is incorrect and building it from source would solve the problem. I have therefore deleted the whole /home/linuxbrew/.linuxbrew directory and restarted the whole process except I used `brew install postgis --build-from-source`. At the end it didn't make any difference, PostGIS currently is still unusable under Linuxbrew since the /home/linuxbrew/.linuxbrew/share/postgresql/extension/postgis.control file is missing.

Please find the result of `brew gist-logs postgis` [here](https://gist.github.com/af4d542b71291f3b618aca7699cb26ea).

**Comment by sandrasi** [comment]
I've found the culprit. The PostGIS formula uses Ruby's `Dir` class to find all the files that need to be installed after the binaries are built: https://github.com/Homebrew/homebrew-core/blob/30f84ab58cbce16b63ba394590c1e9c836d3b32b/Formula/postgis.rb#L55-L61

The `Dir['some/global/**/pattern/*']`, however, won't match on UNIX hidden files (those of which name starts with a dot). Since Linuxbrew is installed under _/home/linuxbrew/.linuxbrew_ by default, the temporary directory where PostGIS is built to also contains this path and therefore `Dir` ignores everything (this explains the warnings listed above in the description). This is not a problem on a MacBook because Homebrew is installed to `/usr/local/Homebrew` (or so I believe) so no UNIX hidden files are involved.

I've fixed it locally on my machine by adding this method to the formula (_/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/postgis.rb_):
```ruby
class Postgis < Formula
  ...
  def dir(pattern)
    Dir.glob(pattern, File::FNM_DOTMATCH).select do |path|
      !path.end_with?('/.', '/..')
    end
  end
  ...
end
```
and then replaced every `Dir["some/global/**/pattern/*"]` call with `dir("some/global/**/pattern/*")`, and finally reinstalled PostGIS (`$ brew reinstall postgis`).

What it does is that it
- matches on _all_ files, including those whose name starts with a dot, and
- filters out the current (`.`) and the parent (`..`) directories, which are also returned by this version of `Dir` (and would probably generate some other error during installation).

I'll make a PR with this proposed change and see if it gets approved. Until then - if anybody has the same problem installing PostGIS in Linuxbrew - this comment offers a temporary workaround.
